File: Core/src/main/java/com/inverce/mod/core/verification/Conditions.java
package com.inverce.mod.core.verification;

import android.database.Cursor;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;

import java.lang.ref.WeakReference;
import java.util.Collection;
import java.util.Map;

@SuppressWarnings("unused")
public class Conditions {
    private Conditions() { }

    @SafeVarargs
    public static <T> T firstNonNull(@Nullable T value, @NonNull T ... values) {
        if (value != null) return value;
        for (T k: values) {
            if (k != null) {
                return k;
            }
        }
        return null;
    }

    public static boolean notNullOrEmpty(@Nullable String value) {
        return value != null && value.length() > 0;
    }

    public static boolean notNullOrEmpty(@Nullable WeakReference<?> value) {
        return value != null && value.get() != null;
    }

    public static <T> boolean notNullOrEmpty(@Nullable T [] value) {
        return value != null && value.length > 0;
    }

    public static boolean notNullOrEmpty(@Nullable Collection<?> value) {
        return value != null && value.size() > 0;
    }

    public static boolean notNullOrEmpty(@Nullable Cursor value) {
        return value != null && value.getCount() > 0;
    }

    public static boolean nullOrEmpty(@Nullable String value) {
        return value == null || value.length() == 0;
    }

    public static boolean nullOrEmpty(@Nullable WeakReference<?> value) {
        return value == null || value.get() == null;
    }

    public static <T> boolean nullOrEmpty(@Nullable T [] value) {
        return value == null || value.length == 0;
    }

    public static boolean nullOrEmpty(@Nullable Collection<?> value) {
        return value == null || value.size() == 0;
    }

    public static boolean nullOrEmpty(@Nullable Map<?, ?> value) {
        return value == null || value.size() == 0;
    }

    public static boolean nullOrEmpty(@Nullable Cursor value) {
        return value == null || value.getCount() == 0;
    }

    public static boolean isVisible(@Nullable View view) {
        return view != null && view.getVisibility() == View.VISIBLE;
    }
}

File: Core/src/main/java/com/inverce/mod/v2/core/verification/Conditions.kt
@file:JvmName("Conditions")

package com.inverce.mod.v2.core.verification

import android.database.Cursor
import java.lang.ref.WeakReference

@SafeVarargs
fun <T> firstNonNull(value: T?, vararg values: T) = when (value) {
    null -> value
    else -> values.firstOrNull { it != null }
}

fun noneNull(vararg objects: Any?) = objects.none { it == null }

fun String?.isNotNullOrEmpty() = this?.isNotEmpty() ?: false
fun WeakReference<*>?.isNotNullOrEmpty() = this?.get() != null
fun <T> Array<T>?.isNotNullOrEmpty() = this?.isNotEmpty() ?: false
fun Collection<*>?.isNotNullOrEmpty() = this?.isNotEmpty() ?: false
fun Cursor?.isNotNullOrEmpty() = this?.count ?: 0 > 0

fun String?.isNullOrEmpty() = this?.isEmpty() ?: true
fun WeakReference<*>?.isNullOrEmpty() = this?.get() == null ?: true
fun <T> Array<T>?.isNullOrEmpty() = this?.isEmpty() ?: true
fun Collection<*>?.isNullOrEmpty() = this?.isEmpty() ?: true
fun Map<*, *>?.isNullOrEmpty() = this?.isEmpty() ?: true
fun Cursor?.isNullOrEmpty() = this?.count

internal inline fun <T, Y> nullPass(value: T?, pass: (T) -> Y?): Y? = when (value) {
    null -> null
    else -> pass(value)
}

fun <T> T?.nullWhen(filter: T) = nullWhen { this == filter }
inline fun <T> T?.nullWhen(filter: T.() -> Boolean) = when {
    this == null -> null
    filter(this) -> null
    else -> this
}